llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp

```diff
@@ -0,0 +1,393 @@
+    if (!Edges.count(T))
+      continue;
+
+    for (const Type *TT : Edges.at(T)) {
+      if (!Visited.count(TT))
+        queue.push(TT);
+    }
+  }
+}
+
+/// Propagates given aspects to all types in module @M. Function accepts
+/// aspects in @TypesWithAspects reference and writes a result in this
+/// reference.
+/// Type T in the result contains an aspect A if Type T is a composite
+/// type (array, struct, vector) which contains elements/fields of
+/// another type TT, which in turn uses the aspect A.
+/// @TypesWithAspects argument consist of known types with aspects
+/// from metadata information.
+///
+/// The algorithm is the following:
+/// 1) Make a list of all structure types from module @M. The list also
+///    contains DoubleTy since it is optional as well.
+/// 2) Make from list a type graph which consist of nodes corresponding to types
+///    and directed edges between nodes. An edge from type A to type B
+///    corresponds to the fact that A is contained within B.
+///    Examples: B is a pointer to A, B is a struct containing field of type A.
+/// 3) For every known type with aspects propagate it's aspects over graph.
+///    Every propagation is a separate run of BFS algorithm.
+///
+/// Time complexity: O((V + E) * T) where T is the number of input types
+/// containing aspects.
+void PropagateAspectsToOtherTypesInModule(
+    const Module &M, TypeToAspectsMapTy &TypesWithAspects) {
+  std::unordered_set<const Type *> TypesToProcess;
+  Type *DoubleTy = Type::getDoubleTy(M.getContext());
+
+  // 6 is taken from sycl/include/CL/sycl/aspects.hpp
+  static constexpr int AspectFP64 = 6;
+  TypesWithAspects[DoubleTy].insert(AspectFP64);
+
+  TypesToProcess.insert(DoubleTy);
+  for (Type *T : M.getIdentifiedStructTypes())
+    TypesToProcess.insert(T);
+
+  TypesEdgesTy Edges;
+  for (const Type *T : TypesToProcess) {
+    for (const Type *TT : T->subtypes()) {
+      // If TT = %A*** then we want to get TT = %A
+      while (TT->isPointerTy())
+        TT = TT->getContainedType(0);
+
+      // We are not interested in some types. For example, IntTy.
+      if (!TypesToProcess.count(TT))
+        continue;
+
+      Edges[TT].push_back(T);
+    }
+  }
+
+  TypeToAspectsMapTy Result;
+  for (const Type *T : TypesToProcess) {
+    AspectsSetTy &Aspects = TypesWithAspects[T];
+    PropagateAspectsThroughTypes(Edges, T, Aspects, TypesWithAspects);
+  }
+}
+
+/// Returns all aspects which might be reached from type @T.
+/// It encompases composite structures and pointers.
+/// NB! This function inserts new records in @Types map for new discovered
+/// types. For the best perfomance pass this map in the next invocations.
+AspectsSetTy GetAspectsFromType(const Type *T, TypeToAspectsMapTy &Types) {
+  auto It = Types.find(T);
+  if (It != Types.end())
+    return It->second;
+
+  // This is essential to no get into infinite recursive loops.
+  Types[T] = {};
+  AspectsSetTy Result;
+
+  for (const Type *TT : T->subtypes()) {
```

Comment:
Why we collect aspects through tree again? Shouldn't Types[T] already contain all aspects for it and its subtypes after `PropagateAspectsThroughTypes`?

Reply:
`PropagateAspectsThroughTypes` tries to traverse over Identified types like structs mostly. While `getAspectsFromType` gets arguments from instructions which may operate with non-identified types. For example, there can be
`bitcast i64 %a to %T***`. Here, we want to get `%T` instead of `%T***`. The same with arrays and vectors.